src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs

```diff
@@ -166,6 +166,18 @@ async Task OnBind(ListenOptions options, CancellationToken onBindCancellationTok
                     // Http/1 without TLS, no-op HTTP/2 and 3.
                     if (hasHttp1)
                     {
+                        if (options.ProtocolsSetExplicitly)
+                        {
+                            if (hasHttp2)
+                            {
+                                Trace.LogWarning(CoreStrings.Http2DisabledWithHttp1AndNoTls, options.EndPoint);
```

Comment:
Usually, logging is done using defined logs. See KestrelTrace. I'm not sure why logging in this file is using LogWarning/LogInfo/etc directly.

Reply:
I'm not sure what "defined logs" are - maybe they're more structured or have corresponding ETW events?

Reply:
Do we localize those messages?  If so, how?

Reply:
See KestrelTrace - https://github.com/dotnet/aspnetcore/blob/9114f148b0ec923ebad046207598fcfa092f9d4e/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelTrace.General.cs

Most logging hangs off it.

We put log messages in resource files, but as far as I know, we don't localize.

Reply:
To clarify, I understand that you're suggesting I use `KestrelTrace` and `LoggerMessageAttribute`, but I'm wondering what all that implies and whether "defined logs" is a more widely used term I should be aware of.

Reply:
"defined logs" refers to the `LoggerMessage.Define` way of logging
https://learn.microsoft.com/dotnet/core/extensions/high-performance-logging#define-a-logger-message

Reply:
We don't use "trace." anywhere, we use Microsoft.Extensions.Logging. @JamesNK is pointing you at the existing logs so you can add more to that.

Reply:
Oh, Trace here is KestrelTrace...

Reply:
Added the messages to `KestrelTrace.GeneralLog`.